### DISCUSSION TITLE
Background Navigation Service no option "Always allow" use location

### DISCUSSION THREAD

**Discussion by NickSlash** [osmandapp/OsmAnd#10912] [Q&A]
### Description
One of the recent updates has caused navigation to be a lot less responsive and there is a message ("the background navigation service requires a location provider to be turned on") that pops up frequently during navigation. 

I had a look at the permissions for the app and there are two options, "deny" and "allow when in use". I cant be certain but I think that previously there was an "allow" option.

### How to reproduce?
start navigation to any destination

### Your Environment
OsmAnd Version: osmand+ 3.9.5 2021-02-03
Android/iOS version: android 10
Device model: oneplus nord 

**Maps used (online or offline):**   offline map
If you have an issue related to offline maps, tell us the exact name of the map file where the issue occurs and its edition date.

**Comment by vshcherb** [comment]
Yes, we fixed couple of issues in 3.9.6 - and there is no "Always allow" this is a restrictions from Google Play but the app should be working fine now.

**Reply by Krandelbord** [reply]
> Yes, we fixed couple of issues in 3.9.6 - and there is no "Always allow" this is a restrictions from Google Play but the app should be working fine now.

So how can I record track with screen off?
On my Xperia 5 IV, turning off screen stops track recording.

**Reply by scaidermern** [reply]
Check https://docs.osmand.net/docs/user/troubleshooting/track-recording-issues and https://dontkillmyapp.com/

**Comment by sakhnik** [comment]
@vshcherb, I remember "Always allow" was there, and I had to set it to be able to navigate or record tracks with the screen off. Will it still work without "Always allow" or are there new limitations?

**Comment by vshcherb** [comment]
https://docs.osmand.net/en/main@latest/osmand/troubleshooting/track-recording-issues update documentation here about always allow

**Reply by scaidermern** [reply]
Link is outdated, documentation is now here: https://docs.osmand.net/docs/user/troubleshooting/track-recording-issues

**Comment by jfburke3** [comment]
Well, something isn't right with the Play store rules, or the Android API documentation, because other apps have "Allow all the time" permission for Location in Android 10.  I don't know how they're able to get away with it but y'all can't.

**Reply by ohan-se** [reply]
Hi!
This might have something to do with it:

According to https://medium.com/@ty2/understanding-permissions-for-background-location-on-android-11-and-below-bc3ad9be320a 

"How permissions behaves on:
* Android 9 (and below), 
* Android 10 and 
* Android 11 (and above) 
for background locations are all different."

Further:

"For Android 10 and up, you have to explicitly request for background location. So just like you requested for Manifest.permission.ACCESS_FINE_LOCATION, you now do so for Manifest.permission.ACCESS_BACKGROUND_LOCATION."

Best regards!

**Reply by download13** [reply]
What are the rules on requesting that access? This app definitely has a good reason to request it.

I was riding my bicycle to a park with an earbud in to hear the directions, and it kept saying it had lost location lock or something.
It would regain it once I stopped, pulled out my phone and messed with the app for a bit, but that's kind of a ridiculous for an app that provides audio directions.

It's odd because for plenty of other apps I can go into their permissions and allow precise location all the time instead of just when using the app, but OsmAnd doesn't have that option at all.

**Reply by scaidermern** [reply]
> I was riding my bicycle to a park with an earbud in to hear the directions, and it kept saying it had lost location lock or something. It would regain it once I stopped, pulled out my phone and messed with the app for a bit, but that's kind of a ridiculous for an app that provides audio directions.

This shouldn't happen. Try to disable battery optimization for OsmAnd in your Android settings.

Also see https://dontkillmyapp.com/

**Reply by ohan-se** [reply]
Update: I can recommend the app "GPS connected". 

It is listed in OsmAnd as one of the possible options for a shortcut that can be reached by tappig the "GPS info" in the right panel. 

https://play.google.com/store/apps/details?id=org.bruxo.gpsconnected

"GPS connected is a separate app that both: 
A) let you to easily toggle to force the GPS to stay connected, and 
B) displays in a elegant notification: 
* the accuracy of the GPS position and 
* the number of connected satellites. 

Brilliant! 
Try it for free! 
Then: the pro is not expensive, and it's with giving the creator a buck to remove the ads!

I primerly use it for ceeping GPS from dropping connections for another app I run in the background, that's sending GPS info via Bluetooth to my digital system camera.

**Reply by download13** [reply]
@scaidermern Nevermind, it must have been an issue with MIUI. I swapped to a different OS and it works fine now.

**Reply by jankytech** [reply]
I am also having this issue.
One Plus 7 Pro
Android 11
OSMAnd v4.1.11
Battery saver is set to "do not optimise" for this app
![Screenshot_20220707-211651](https://user-images.githubusercontent.com/13768749/177864029-b8dfbfdb-ee40-47d3-8e28-9f76d6b25769.jpg)

**Reply by AntiSol** [reply]
Agreed, I see the same thing on android 10, other apps have "allow all the time" as an option, but not osmand. And as soon as I switch away from osmand it loses its gps fix, and then after I switch back I have to wait for it to re-establish its fix.

**Reply by scaidermern** [reply]
> And as soon as I switch away from osmand it loses its gps fix, and then after I switch back I have to wait for it to re-establish its fix.

Although you disabled battery optimization for OsmAnd?

**Reply by AntiSol** [reply]
yes, I've disabled battery optimisation entirely.

it seems to me that [this](https://github.com/osmandapp/OsmAnd/discussions/10912#discussioncomment-975004) is the thing.

**Reply by scaidermern** [reply]
> yes, I've disabled battery optimisation entirely.
> 
> it seems to me that [this](https://github.com/osmandapp/OsmAnd/discussions/10912#discussioncomment-975004) is the thing.

I don't think so. I've used OsmAnd with Android 10 (LineageOS) for more than 4 years and never had problems with track recording although I'm using it regularly. Except for track recording with energy saving enabled, this doesn't work and is a known problem (and maybe unfixable).

Also check your device at https://dontkillmyapp.com/

**Reply by AntiSol** [reply]
I've already been through dontkillmyapp.com.

also note that linageos has a bunch of differences from stock android.

It seems you've missed a detail in my report. I am not talking about track recording. I am talking about just using the app. If I switch away from the app for any time at all, when I come back to osmand it says it has 0 satellites and I need to wait for it to re-establish a fix.

To reproduce:
1. start osmand
2. wait for a fix. don't record your track.
3. switch to another app for <5 seconds
4. switch back to osmand
5. observe no fix and 0 satellites in use
6. wait for fix again (should be faster this time, but still, not instant. If the app had "always allow gps" permission the fix would not be lost in the first place)

**Reply by scaidermern** [reply]
That's normal behavior. Why should it drain your battery in the background if it is unclear whether you need again or not? If you want OsmAnd to stay active, start a navigation or a track recording.

**Reply by AntiSol** [reply]
> That's normal behavior.

That's not what happens on my old lineageos14 device.

> if it is unclear whether you need again or not?

It's not unclear at all: If I don't need it, I will close it, which is a distinct operation from switching away to another app for <5 seconds.

> If you want OsmAnd to stay active, start a navigation or a track recording.

...even though I have no need to record my track, and am not navigating anywhere...
